src/parsers/manifest/dash/wasm-parser/ts/dash-wasm-parser.ts

```diff
@@ -0,0 +1,485 @@
+  private _parsersStack : ParsersStack;
+  /**
+   * Created WebAssembly instance.
+   * `null` when no WebAssembly instance is created.
+   */
+  private _instance : WebAssembly.WebAssemblyInstantiatedSource | null;
+  /**
+   * Information about the data read by the DASH-WASM parser.
+   * `null` if we're not parsing anything for the moment.
+   */
+  private _mpdData : {
+    /**
+     * First not-yet read position in `mpd`, in bytes.
+     * When the parser asks for new data, we start giving data from that point
+     * on.
+     */
+    cursor : number;
+    /**
+     * Complete data that needs to be parsed.
+     * This is either the full MPD or xlinks.
+     */
+    mpd : ArrayBuffer;
+  } | null;
+  /**
+   * Warnings event currently encountered during parsing.
+   * Emptied when no parsing is taking place.
+   */
+  private _warnings : Error[];
+  /**
+   * Memory used by the WebAssembly instance.
+   * `null` when no WebAssembly instance is created.
+   */
+  private _linearMemory : WebAssembly.Memory | null;
+
+  /**
+   * `true` if we're currently parsing a MPD.
+   * Used to explicitely forbid parsing a MPD when a parsing operation is
+   * already pending, as the current logic cannot handle that.
+   *
+   * Note that this is not needed for now because parsing should happen
+   * synchronously.
+   * Still, this property was added to make it much more explicit, in case of
+   * future improvements.
+   */
+  private _isParsing : boolean;
+  /**
+   * Create a new `DashWasmParser`.
+   * @param {object} opts
+   */
+  constructor() {
+    this._parsersStack = new ParsersStack();
+    this._instance = null;
+    this._mpdData = null;
+    this._linearMemory = null;
+    this.status = "uninitialized";
+    this._initProm = null;
+    this._warnings = [];
+    this._isParsing = false;
+  }
+
+  /**
+   * Returns Promise that will resolve when the initialization has ended (either
+   * with success, in which cases the Promise resolves, either with failure, in
+   * which case it rejects the corresponding error).
+   *
+   * This is actually the exact same Promise than the one returned by the first
+   * `initialize` call.
+   *
+   * If that method was never called, returns a rejecting Promise.
+   * @returns {Promise}
+   */
+  public waitForInitialization() : Promise<void> {
+    return this._initProm ??
+           PPromise.reject("No initialization performed yet.");
+  }
+
+  public async initialize(opts : IDashWasmParserOptions) : Promise<void> {
+    if (this.status !== "uninitialized") {
+      return PPromise.reject(new Error("DashWasmParser already initialized."));
+    } else if (!this.isCompatible()) {
+      this.status = "failure";
+      return PPromise.reject(new Error("Target not compatible with WebAssembly."));
+    }
+    this.status = "initializing";
+
+    const parsersStack = this._parsersStack;
+
+    /** Re-used TextDecoder instance. */
+    const textDecoder = new TextDecoder();
```

Comment:
Maybe we want, as we do it in other parts of the RxPlayer, to check if TextDecoder is available and use something else if it isn't, no ?

Reply:
I just noticed that you check if it is available in the isCompatible function. Is it really what we want ?

Reply:
Even if it seems to be always available when WebAssembly is available, it seems cleaner to check for it, no ?
It could not be the case anymore in the future (even if it seems to be an unusual case).